File: backend/src/main/java/com/travelhub/service/AuthService.java
package com.travelhub.service;

import com.travelhub.dto.AuthRequest;
import com.travelhub.dto.AuthResponse;
import com.travelhub.dto.RegisterRequest;
import com.travelhub.dto.updateAuth;
import com.travelhub.entity.OtpVerification;
import com.travelhub.entity.User;
import com.travelhub.repository.UserRepository;
import com.travelhub.repository.OtpVerificationRepository;
import com.travelhub.security.JwtUtil;
import lombok.RequiredArgsConstructor;

import java.time.LocalDateTime;
import java.util.Random;

import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

@Service
@RequiredArgsConstructor
public class AuthService {
    
    // DEMO MODE: Set to false for production OTP verification
    private static final boolean DEMO_MODE = true;
    
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil jwtUtil;
    private final AuthenticationManager authenticationManager;
    private final OtpVerificationRepository otpVerificationRepository;
    private final EmailOtpService emailOtpService;
    private final CloudinaryService cloudinaryService;
    
    private void verifyOtp(String phone, String otp) {
        // DEMO MODE: Accept any OTP for demo purposes
        if (DEMO_MODE) {
            return; // Accept any OTP in demo mode
        }

        // Production OTP verification
        OtpVerification otpVerification = otpVerificationRepository
                .findTopByPhoneOrderByExpiresAtDesc(phone)
                .orElseThrow(() -> new RuntimeException("OTP not found"));

        if (otpVerification.getExpiresAt().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("OTP expired");
        }

        if (!otpVerification.getOtp().equals(otp)) {
            throw new RuntimeException("Invalid OTP");
        }

        otpVerification.setVerified(true);
        otpVerificationRepository.save(otpVerification);
    }

    public void otpSending(Long phone) {

        String otp = String.valueOf(100000 + new Random().nextInt(900000));
        
        OtpVerification otpVerification = OtpVerification.builder()
                .phone(String.valueOf(phone))
                .otp(otp)
                .verified(false)
                .expiresAt(LocalDateTime.now().plusMinutes(5))
                .build();

        otpVerificationRepository.save(otpVerification);
        // TODO: integrate SMS provider (Twilio, MSG91, etc.)
    }

    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already exists");
        }
        if (userRepository.existsByPhone(request.getPhone())) {
            throw new RuntimeException("Number already exists");
        }
        
        verifyOtp(request.getPhone(), request.getOtp());
        
        User user = User.builder()
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .fullName(request.getFullName())
                .phone(request.getPhone())
                .whatsappNumber(request.getWhatsappNumber())
                .phoneVerified(Boolean.TRUE)
                .emailVerified(Boolean.FALSE)
                .build();
        
        userRepository.save(user);
        
        String token = jwtUtil.generateToken(user.getEmail(), user.getPhone());
        
        return AuthResponse.builder()
                .token(token)
                .userId(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .phone(user.getPhone())
                .whatsappNumber(user.getWhatsappNumber())
                .build();
    }
    
    public AuthResponse login(AuthRequest request) {
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getPhone(), request.getPassword())
        );
        
        User user = userRepository.findByPhone(request.getPhone())
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        String token = jwtUtil.generateToken(user.getEmail(), user.getPhone());
        
        return AuthResponse.builder()
                .token(token)
                .userId(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .phone(user.getPhone())
                .whatsappNumber(user.getWhatsappNumber())
                .build();
    }
    
    public User getCurrentUser(String phone) {
        return userRepository.findByPhone(phone)
                .orElseThrow(() -> new RuntimeException("User not found"));
    }

    public User updateCurrentUser(String phone, updateAuth updateAuth) {
        User user = userRepository.findByPhone(phone)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // ===== ALLOWED UPDATES ONLY =====
        if (updateAuth.getFullName() != null && !updateAuth.getFullName().isBlank()) {
            user.setFullName(updateAuth.getFullName());
        }
        if (updateAuth.getEmail() != null && !updateAuth.getEmail().isBlank()) {
            String newEmail = updateAuth.getEmail().trim();
            // Only update if actually changed
            if (!newEmail.equalsIgnoreCase(user.getEmail())) {
                // Check uniqueness
                if (userRepository.existsByEmail(newEmail)) {
                    throw new RuntimeException("Email already in use by another account");
                }
                user.setEmail(newEmail);
                user.setEmailVerified(Boolean.FALSE); // reset verification on email change
            }
        }
        if (updateAuth.getWhatsappNumber() != null && !updateAuth.getWhatsappNumber().isBlank()) {
            user.setWhatsappNumber(updateAuth.getWhatsappNumber());
        }
        if (updateAuth.getBio() != null) {
            String bio = updateAuth.getBio().trim();
            user.setBio(bio.isEmpty() ? null : bio);
        }
        // DO NOT allow updates for:
        // password, rating, reviewCount, numberOfTrips
        return userRepository.save(user);
    }

    public User getUserById(Long id) {
        return userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));
    }

    public void sendOtpForRegistration(String phone) {
        // For now, always send OTP and return success
        // TODO: Later add validation to check if phone already exists
        // TODO: Integrate SMS provider to actually send OTP
        
        try {
            // Send OTP for registration
            otpSending(Long.parseLong(phone));
        } catch (Exception e) {
            // For demo purposes, always return success even if there's an error
            // TODO: Remove this catch block and add proper error handling later
            System.out.println("Error sending OTP (ignored for demo): " + e.getMessage());
        }
        
        // TODO: Send SMS with OTP
        // For now, OTP is saved in database and can be retrieved for testing
    }

    public void forgotPassword(String phone) {
        // Check if user exists
        if (!userRepository.existsByPhone(phone)) {
            throw new RuntimeException("User not found with this phone number");
        }
        
        // Send OTP for password reset
        otpSending(Long.parseLong(phone));
        
        // TODO: Send SMS with OTP
        // For now, OTP is saved in database and can be retrieved for testing
    }

    public void resetPassword(String phone, String otp, String newPassword) {
        // Verify OTP
        verifyOtp(phone, otp);
        
        // Find user
        User user = userRepository.findByPhone(phone)
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        // Update password
        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }

    public void sendOtpForEmailVerification(String phone, String email) {
        User user = userRepository.findByPhone(phone)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (email == null || email.isBlank()) {
            throw new RuntimeException("Email is required");
        }
        if (!user.getEmail().equalsIgnoreCase(email.trim())) {
            throw new RuntimeException("Email does not match your account email");
        }

        emailOtpService.sendEmailOtp(user, user.getEmail());
    }

    public void verifyEmailOtp(String phone, String otp) {
        User user = userRepository.findByPhone(phone)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (otp == null || otp.isBlank()) {
            throw new RuntimeException("Otp is required");
        }
        boolean ok = emailOtpService.verifyEmailOtp(user, user.getEmail(), otp.trim());
        if (!ok) {
            throw new RuntimeException("Invalid or expired OTP");
        }

        user.setEmailVerified(Boolean.TRUE);
        userRepository.save(user);
    }

    public User uploadProfilePhoto(String phone, MultipartFile photo) {
        User user = userRepository.findByPhone(phone)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (photo == null || photo.isEmpty()) {
            throw new RuntimeException("Photo is required");
        }

        String url = cloudinaryService.uploadMedia(photo, user.getId() + "/profile");
        user.setProfilePhoto(url);
        return userRepository.save(user);
    }
}

File: backendNode/src/services/auth.service.ts
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";
import { prisma } from "../lib/prisma.js";
import { redis } from "../lib/redis.js";
import { env } from "../config/env.js";
import { ApiError } from "../utils/http.js";
import { userPublic } from "../domain/mappers.js";
import { hasMailConfig, mailer } from "../lib/mailer.js";
import { hasCloudinaryConfig, cloudinary } from "../lib/cloudinary.js";

const DEMO_MODE = true;

type RegisterInput = {
  email: string;
  password: string;
  fullName: string;
  phone: string;
  Otp: string;
  whatsappNumber?: string;
};

export function signToken(email: string, phone: string): string {
  return jwt.sign({ Email: email }, env.jwtSecret, {
    subject: phone,
    expiresIn: Math.floor(env.jwtExpirationMs / 1000),
  });
}

async function verifyPhoneOtp(phone: string, otp: string): Promise<void> {
  if (DEMO_MODE) return;
  const record = await prisma.otpVerification.findFirst({
    where: { phone },
    orderBy: { expiresAt: "desc" },
  });
  if (!record) throw new ApiError(400, "OTP not found");
  if (record.expiresAt < new Date()) throw new ApiError(400, "OTP expired");
  if (record.otp !== otp) throw new ApiError(400, "Invalid OTP");
  await prisma.otpVerification.update({
    where: { id: record.id },
    data: { verified: true },
  });
}

export async function register(input: RegisterInput) {
  const email = input.email.trim().toLowerCase();
  const phone = input.phone.trim();
  const existingEmail = await prisma.user.findUnique({ where: { email } });
  if (existingEmail) throw new ApiError(400, "Email already exists");
  const existingPhone = await prisma.user.findUnique({ where: { phone } });
  if (existingPhone) throw new ApiError(400, "Number already exists");

  await verifyPhoneOtp(phone, input.Otp.trim());
  const passwordHash = await bcrypt.hash(input.password, 10);

  const user = await prisma.user.create({
    data: {
      email,
      password: passwordHash,
      fullName: input.fullName.trim(),
      phone,
      whatsappNumber: input.whatsappNumber?.trim() || null,
      phoneVerified: true,
      emailVerified: false,
    },
  });

  return {
    token: signToken(user.email, user.phone),
    userId: Number(user.id),
    email: user.email,
    fullName: user.fullName,
    phone: user.phone,
    whatsappNumber: user.whatsappNumber,
    role: "USER",
  };
}

export async function login(phone: string, password: string) {
  const user = await prisma.user.findUnique({ where: { phone } });
  if (!user) throw new ApiError(401, "Invalid email or password");
  const ok = await bcrypt.compare(password, user.password);
  if (!ok) throw new ApiError(401, "Invalid email or password");

  return {
    token: signToken(user.email, user.phone),
    userId: Number(user.id),
    email: user.email,
    fullName: user.fullName,
    phone: user.phone,
    whatsappNumber: user.whatsappNumber,
    role: "USER",
  };
}

export async function getCurrentUser(phone: string) {
  const user = await prisma.user.findUnique({ where: { phone } });
  if (!user) throw new ApiError(400, "User not found");
  return userPublic(user);
}

export async function getUserById(id: number) {
  const user = await prisma.user.findUnique({ where: { id: BigInt(id) } });
  if (!user) throw new ApiError(400, "User not found");
  const payload = userPublic(user);
  return {
    id: payload.id,
    fullName: payload.fullName,
    phone: payload.phone,
    whatsappNumber: payload.whatsappNumber,
    bio: payload.bio,
    rating: payload.rating,
    reviewCount: payload.reviewCount,
    numberOfTrips: payload.numberOfTrips,
    profilePhoto: payload.profilePhoto,
    phoneVerified: payload.phoneVerified,
    emailVerified: payload.emailVerified,
  };
}

export async function updateCurrentUser(
  phone: string,
  updates: { fullName?: string; email?: string; whatsappNumber?: string; bio?: string },
) {
  const user = await prisma.user.findUnique({ where: { phone } });
  if (!user) throw new ApiError(400, "User not found");

  const patch: Record<string, unknown> = {};
  if (updates.fullName?.trim()) patch.fullName = updates.fullName.trim();
  if (updates.whatsappNumber?.trim()) patch.whatsappNumber = updates.whatsappNumber.trim();
  if (updates.bio !== undefined) patch.bio = updates.bio.trim() || null;
  if (updates.email?.trim()) {
    const email = updates.email.trim().toLowerCase();
    if (email !== user.email.toLowerCase()) {
      const exists = await prisma.user.findUnique({ where: { email } });
      if (exists) throw new ApiError(400, "Email already in use by another account");
      patch.email = email;
      patch.emailVerified = false;
    }
  }

  const saved = await prisma.user.update({
    where: { id: user.id },
    data: patch,
  });
  return userPublic(saved);
}

export async function sendOtpForRegistration(phone: string) {
  const otp = String(100000 + Math.floor(Math.random() * 900000));
  await prisma.otpVerification.create({
    data: {
      phone,
      otp,
      verified: false,
      expiresAt: new Date(Date.now() + 5 * 60 * 1000),
    },
  });
}

export async function forgotPassword(phone: string) {
  const user = await prisma.user.findUnique({ where: { phone } });
  if (!user) throw new ApiError(400, "User not found with this phone number");
  await sendOtpForRegistration(phone);
}

export async function resetPassword(phone: string, otp: string, newPassword: string) {
  await verifyPhoneOtp(phone, otp);
  const user = await prisma.user.findUnique({ where: { phone } });
  if (!user) throw new ApiError(400, "User not found");
  const hash = await bcrypt.hash(newPassword, 10);
  await prisma.user.update({
    where: { id: user.id },
    data: { password: hash },
  });
}

function otpKey(fullName: string, email: string): string {
  return `otp:email:${fullName.trim().toLowerCase().replace(/\s+/g, "_")}_${email.trim().toLowerCase()}`;
}

export async function sendEmailOtp(phone: string, email: string) {
  const user = await prisma.user.findUnique({ where: { phone } });
  if (!user) throw new ApiError(400, "User not found");
  const normalized = email.trim().toLowerCase();
  if (normalized !== user.email.toLowerCase()) {
    throw new ApiError(400, "Email does not match your account email");
  }

  const otp = String(100000 + Math.floor(Math.random() * 900000));
  await redis.connect().catch(() => undefined);
  await redis.set(otpKey(user.fullName, normalized), otp, "EX", 300).catch(() => undefined);

  if (hasMailConfig()) {
    await mailer.sendMail({
      from: env.mailFrom,
      to: normalized,
      subject: "PoolMyTrips - Email Verification OTP",
      text: `Your PoolMyTrips email verification OTP is: ${otp}\n\nThis OTP will expire in 5 minutes.`,
    });
  }
}

export async function verifyEmailOtp(phone: string, otp: string) {
  const user = await prisma.user.findUnique({ where: { phone } });
  if (!user) throw new ApiError(400, "User not found");
  await redis.connect().catch(() => undefined);
  const key = otpKey(user.fullName, user.email);
  const expected = await redis.get(key).catch(() => null);
  if (!expected || expected !== otp.trim()) throw new ApiError(400, "Invalid or expired OTP");
  await redis.del(key).catch(() => undefined);
  await prisma.user.update({
    where: { id: user.id },
    data: { emailVerified: true },
  });
}

export async function uploadProfilePhoto(phone: string, file?: Express.Multer.File) {
  const user = await prisma.user.findUnique({ where: { phone } });
  if (!user) throw new ApiError(400, "User not found");
  if (!file) throw new ApiError(400, "Photo is required");
  if (!hasCloudinaryConfig()) {
    throw new ApiError(400, "Cloudinary is not configured");
  }

  const uploaded = await cloudinary.uploader.upload(file.path, {
    folder: `poolmytrips/${user.id.toString()}/profile`,
    resource_type: "auto",
  });
  const updated = await prisma.user.update({
    where: { id: user.id },
    data: { profilePhoto: uploaded.secure_url },
  });

  return {
    profilePhoto: updated.profilePhoto,
    emailVerified: updated.emailVerified,
    phoneVerified: updated.phoneVerified ?? true,
  };
}
